paddle/fluid/platform/float16_test.cu

```diff
@@ -241,6 +241,43 @@ TEST(float16, lod_tensor_on_gpu) {
   }
 }
 
+template <typename T>
+struct Functor {
+  bool operator()(const T& val) {
+    return std::type_index(typeid(T)) ==
+           std::type_index(typeid(platform::float16));
+  }
+};
+
+TEST(float16, typeid) {
+  // the framework heavily used typeid hash
+  Functor<float16> functor;
+  float16 a = float16(.0f);
+  Functor<int> functor2;
+  int b(0);
+
+  // compile time assert
+  PADDLE_ASSERT(functor(a) == true);
+  PADDLE_ASSERT(functor2(b) == false);
+}
+
+// GPU test
+TEST(float16, isinf) {
+  float16 a;
+  a.x = 0x7c00;
+  float16 b = float16(INFINITY);
+  EXPECT_EQ(std::isinf(a), true);
+  EXPECT_EQ(std::isinf(b), true);
+}
+
+TEST(float16, isnan) {
+  float16 a;
+  a.x = 0x7fff;
```

Comment:
It's certain that `0x7fff` is `nan` but do we need to ensure that numbers like `1e-40` is `nan` too?

Sorry that not familiar with fp16 implement, need some time to look over the design.

Reply:
All the numbers bigger than 0x7fff will be nan. See here. https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/platform/float16.h#L864

Take 1e-40 for example, it will automatically cast into a float16 format through https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/plate form/float16.h#L146
It will be overflow.
https://en.wikipedia.org/wiki/Half-precision_floating-point_format

Reply:
I can add two float number overflow test case.

Reply:
add more overflow test case.